src/language_server_protocol.rs

```diff
@@ -3123,10 +3096,66 @@ impl LanguageClient {
             )?;
         }
 
-        info!("End {}", NOTIFICATION__HandleCursorMoved);
         Ok(())
     }
 
+    fn virtual_texts_from_diagnostics(
+        &self,
+        filename: &str,
+        viewport: viewport::Viewport,
+    ) -> Fallible<Vec<VirtualText>> {
+        let mut virtual_texts = vec![];
+        let diagnostics = self.get(|state| state.diagnostics.clone())?;
+        let diagnosticsDisplay = self.get(|state| state.diagnosticsDisplay.clone())?;
+        let diag_list = diagnostics.get(filename);
+        if let Some(diag_list) = diag_list {
+            for diag in diag_list {
+                if viewport.overlaps(diag.range) {
+                    virtual_texts.push(VirtualText {
+                        line: diag.range.start.line,
```

Comment:
Similarly, `impl From<Diagnostic> for VirtualText`

Reply:
This one is easier than `impl From<CodeLens> for VirtualText` but still, diagnostics virtual texts have their `hl_group` that is obtained from the config of the language server. So the implementation of `From` isn't really possible. Something similar could be possible but it wouldn't be that different than as it is now.